core/src/main/scala/kafka/server/KafkaServer.scala

```diff
@@ -318,8 +318,9 @@ class KafkaServer(
 
         /* start group coordinator */
         // Hardcode Time.SYSTEM for now as some Streams tests fail otherwise, it would be good to fix the underlying issue
-        groupCoordinator = GroupCoordinator(config, zkClient, replicaManager, Time.SYSTEM, metrics)
-        groupCoordinator.startup()
+        groupCoordinator = GroupCoordinator(config, replicaManager, Time.SYSTEM, metrics)
+        groupCoordinator.startup(zkClient.getTopicPartitionCount(Topic.GROUP_METADATA_TOPIC_NAME).
```

Comment:
Why not pass the number of partitions in the constructor instead?

Reply:
Good question.  We won't know what value to provide when using a Raft-based metadata quorum until we have caught up on the metadata log and know our configs.  Providing it in the `startup()` function has no impact in the ZooKeeper world because we invoke that immediately after constructing the instance.  In the Raft-based world, if we provide it in the constructor, it constrains us to not construct the instance until we have caught up and have our configs.  By providing it in the `startup()` method it gives us additional flexibility to potentially instantiate the group coordinator but not yet "start" it.

Reply:
What's the advantage of having a coordinator that is not started?

Reply:
Yeah, it is difficult to justify beyond the fact that `startup()` exists -- maybe it deserves to be gotten rid of.  But that's perhaps a discussion for a separate ticket/PR.

Reply:
The fact that we have stuff like the following implies that the approach taken here isn't so good:
* https://github.com/apache/kafka/pull/10008/files#diff-8bdfc05a89875ddd25c9bd355c102ca2a0556795bb18fba25efc3f9959d789e2R199
* https://github.com/apache/kafka/pull/10008/files#diff-8bdfc05a89875ddd25c9bd355c102ca2a0556795bb18fba25efc3f9959d789e2R188

Reply:
Ok, I reverted and used the same approach as is proposed for the `__transaction_state` partition count in https://github.com/apache/kafka/pull/10013.

Reply:
> What's the advantage of having a coordinator that is not started?

This is a pattern that we are using for a lot of components in the kip-500 branch.  It reflects the fact that in the kip-500 world, we have a "before metadata is available" state and an "after metadata is available" state.  The advantage is that it makes this state explicit rather than being implicit.  When the state is implicit mistakes can easily happen-- for example, someone asks for some piece of metadata that is not available.  The original author of the code may have "just known" this, but it's better to formalize it so that people don't make mistakes.

I think the callback approach is really clunky and tends to obfuscate the code.  What's going on here is really that the number of partitions gets initialized once.  Having a complex callback chain just obscures that and makes it tough to read.